### PR TITLE
[Enhancement] optimize in-predicate with dictionary info

### PR DESCRIPTION
**Pull request by zombee0**
## What I'm doing:
old
perf_q22

new
perf_ssb_in_0830_1

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by dirtysalt — DISMISSED** [review]

File: be/src/storage/column_in_predicate.cpp [hunk 1]
```diff
@@ -365,6 +367,93 @@ class BinaryColumnInPredicate : public ColumnPredicate {
...
+                filter[i - from] = !null_data[i];
+            }
+        }
+
+        for (auto i = from; i < to; i++) {
+            auto index = data[i] >= _bit_mask.size() ? 0 : data[i];
+            filter[i - from] &= _bit_mask[index];
+        }
+
+        for (auto i = from; i < to; i++) {
```
Comment (DorianZheng) [inline be/src/storage/column_in_predicate.cpp:412]: Is it better to merge this for-loop with the one in LOC:407

File: be/src/storage/column_in_predicate.cpp [hunk 2]
```diff
@@ -365,6 +367,93 @@ class BinaryColumnInPredicate : public ColumnPredicate {
...
+        if (column->has_null()) {
+            const NullColumn* null_column = down_cast<const NullableColumn*>(column)->null_column().get();
+            auto null_data = null_column->get_data();
+            for (auto i = from; i < to; i++) {
+                filter[i - from] = !null_data[i];
+            }
+        }
+
+        for (auto i = from; i < to; i++) {
+            auto index = data[i] >= _bit_mask.size() ? 0 : data[i];
```
Comment (DorianZheng) [inline be/src/storage/column_in_predicate.cpp:408]: when does `data[i] >= _bit_mask.size()` happen

File: be/src/storage/column_predicate.h [hunk 3]
```diff
@@ -245,6 +245,9 @@ ColumnPredicate* new_column_cmp_predicate(PredicateType predicate, const TypeInf
 
 ColumnPredicate* new_column_in_predicate(const TypeInfoPtr& type, ColumnId id,
                                          const std::vector<std::string>& operands);
+
+ColumnPredicate* new_dictionary_code_in_predicate(const TypeInfoPtr& type, ColumnId id,
+                                                  const std::vector<int32_t>& operands, size_t size);
 ColumnPredicate* new_column_not_in_predicate(const TypeInfoPtr& type, ColumnId id,
```
Comment (DorianZheng) [inline be/src/storage/column_predicate.h:251]: How about optimize not-in-predicate with dictionary too

File: be/src/storage/column_in_predicate.cpp [hunk 4]
```diff
@@ -365,6 +367,93 @@ class BinaryColumnInPredicate : public ColumnPredicate {
...
+            _bit_mask[item] = 1;
+        }
+    }
+
+    ~DictionaryCodeInPredicate() override = default;
+
+    template <LogicOp Op>
+    inline void t_evaluate(const Column* column, uint8_t* sel, uint16_t from, uint16_t to) const {
+        const Int32Column* dict_code_column;
+        if (column->is_nullable()) {
```
Comment (trueeyu) [inline be/src/storage/column_in_predicate.cpp:389]: Use `ColumnHelper::get_data_column()` instead?

File: be/src/storage/column_in_predicate.cpp (hunk 2 above)
Comment (zombee0) [reply be/src/storage/column_in_predicate.cpp:408]: when append_default, there will be some memory resize_uninitalized

File: be/src/storage/column_predicate.h (hunk 3 above)
Comment (zombee0) [reply be/src/storage/column_predicate.h:251]: maybe next pr? we haven't not-in-predicate in parquet reader now.

File: be/src/storage/column_in_predicate.cpp (hunk 4 above)
Comment (zombee0) [reply be/src/storage/column_in_predicate.cpp:389]: Got it. will do it.

File: be/src/storage/column_in_predicate.cpp (hunk 1 above)
Comment (zombee0) [reply be/src/storage/column_in_predicate.cpp:412]: agree, will do it.

File: be/src/storage/column_in_predicate.cpp [hunk 5]
```diff
@@ -507,4 +590,19 @@ ColumnPredicate* new_column_in_predicate(const TypeInfoPtr& type_info, ColumnId
     }
 }
 
+ColumnPredicate* new_dictionary_code_in_predicate(const TypeInfoPtr& type, ColumnId id,
+                                                  const std::vector<int32_t>& operands, size_t size) {
+    DCHECK(is_integer_type(type->type()));
+    if (operands.size() <= 3 || size > 1024) {
```
Comment (trueeyu) [inline be/src/storage/column_in_predicate.cpp:596]: Why add the check `if (operands.size() <= 3 || size > 1024)`?

**Review by trueeyu — DISMISSED** [review]

**Review by DorianZheng — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**